core/src/main/java/org/apache/calcite/plan/volcano/RelSubset.java

```diff
@@ -98,10 +100,13 @@
   long timestamp;
 
   /**
-   * Flag indicating whether this RelSubset's importance was artificially
-   * boosted.
+   * Physical property state of current subset
+   * 0: logical operators, NONE convention is neither DERIVED nor REQUIRED
+   * 1: traitSet DERIVED from child operators or itself
+   * 2: traitSet REQUIRED from parent operators
+   * 3: both DERIVED and REQUIRED
    */
-  boolean boosted;
+  int state = 0;
```

Comment:
I think `state` should probably be `private`. For the condition in RelSet that checks if `otherSubset.state == 0`, it could be re-written as `!otherSubset.isDerived() && !otherSubset.isRequired()`, or we could add a public shortcut method to check this condition `public boolean isXXX() { return state == 0; }`

Reply:
Fixed, thanks.